### PR TITLE
Try to support non-guest room peek. 

### PR DESCRIPTION
**Pull request by ara4n**
Rename visibility permissions in room settings a bit, and fix how they are persisted.

### CODE REVIEW

**Comment by ara4n** [comment]
@kegsay - does this look like it's going remotely in the right direction?  See also https://github.com/vector-im/vector-web/pull/668

File: src/components/structures/MatrixChat.js [hunk 1]
```diff
@@ -304,6 +305,9 @@ module.exports = React.createClass({
                 });
                 break;
             case 'view_room':
+                // by default we autoPeek rooms, unless we were called explicitly with
+                // autoPeek=false by something like RoomDirectory who has already peeked
+                this.setState({ autoPeek : payload.auto_peek === false ? false : true });
```
Comment (Kegsay) [inline src/components/structures/MatrixChat.js]: This conditional makes no sense. I assume you're trying to catch and cast undefined / null. Just do `Boolean(payload.auto_peek)` then. As it stands, you're NOPing because of the triple equals.

Comment (ara4n) [inline src/components/structures/MatrixChat.js]: no, as the comment says, we set state's autoPeek to false if payload's auto_peek is false.  Otherwise we set it to true (even if payload's auto_peek was undefined).

Comment (Kegsay) [inline src/components/structures/MatrixChat.js]: I don't understand why you are not just doing:

```javascript
this.setState({ autoPeek : payload.auto_peek });
```

As that is functionally identical to what you have written here. You're doing `===` so there will be no type coercion for `undefined`.

File: src/components/structures/RoomView.js [hunk 2]
```diff
@@ -111,27 +116,25 @@ module.exports = React.createClass({
...
-                    return;
-                }
-                var guestAccessEvent = peekedRoom.currentState.getStateEvents("m.room.guest_access", "");
-                if (!guestAccessEvent) {
-                    return;
-                }
-                if (guestAccessEvent.getContent().guest_access === "can_join") {
+            if (this.props.autoPeek) {
+                console.log("Attempting to peek into room %s", this.props.roomId);
+                MatrixClientPeg.get().peekInRoom(this.props.roomId).done(() => {
```
Comment (Kegsay) [inline src/components/structures/RoomView.js]: You state in the description for `autoPeekDone` that it "track whether our autoPeek (if any) has completed)". I'm confused what this means in practice. Does this mean if it was *successful*? If it's just a flag for determining whether or not to show the loading spinner, then you need to change the `done()` to `finally()` to catch error conditions.

**Comment by Kegsay** [comment]
LGTM aside from comments. Blocked on #103 landing due to this PR being based off it.

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (ara4n) [inline src/components/structures/MatrixChat.js]: i want:

```
if (payload.auto_peek === undefined) {
    this.setState({ autoPeek : true }); // default to true if not specified
}
else {
    this.setState({ autoPeek: payload.auto_peek });
}
```

which is what the ternary is trying to do more succinctly.  Which of us is on the wrong page? :D

File: src/components/structures/RoomView.js (hunk 2 above)
Comment (ara4n) [inline src/components/structures/RoomView.js]: yes, it means that the autoPeek has completed i.e. finished - whether that was successful or failure.  have switched to finally() in db7d863 - thanks.

File: src/components/structures/RoomView.js [hunk 3]
```diff
@@ -111,27 +116,25 @@ module.exports = React.createClass({
         // We can /peek though. If it fails then we present the join UI. If it
         // succeeds then great, show the preview (but we still may be able to /join!).
         if (!this.state.room) {
-            console.log("Attempting to peek into room %s", this.props.roomId);
-            MatrixClientPeg.get().peekInRoom(this.props.roomId).done(() => {
-                // we don't need to do anything - JS SDK will emit Room events
-                // which will update the UI. We *do* however need to know if we
-                // can join the room so we can fiddle with the UI appropriately.
```
Comment (ara4n) [inline src/components/structures/RoomView.js]: @kegsay - I don't get this.  Why can't we fiddle with the UI when processing the Room Events (as I've changed it to do, handling this state in onNewRoom?)

**Comment by ara4n** [comment]
@Kegsay ptal at my answers

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (Kegsay) [inline src/components/structures/MatrixChat.js]: OH! I understand now. I would c/p that exactly and use that because that is by far the clearer syntax even if it isn't as terse.

Now that I know this, your original line works as you intend it to, it's just I didn't understand your intent from the code. As an aside, `if undefined default true` is generally hard to do clearly and succinctly in Javascript and usually ends up being *the single use case* for `==`:

```javascript
if (someVar == null) { // returns true for undefined and null, but false for the literal false and 0
  someVar = true;
}
// do something with someVar

File: src/components/structures/RoomView.js (hunk 3 above)
Comment (Kegsay) [inline src/components/structures/RoomView.js]: Both work. I had no reason to prefer one or the other. We just need to make sure that if we discover a new room (via peeking or via joining) that we check which screen to show (join UI etc)

**Comment by Kegsay** [comment]
LGTM - Needs #103 merged before landing.

**Comment by Kegsay** [comment]
Not getting any scrollback requests when testing this. The room is blank (sending messages and receiving new ones work though)

**Comment by Kegsay** [comment]
Perf regressions from my invite-search branch due to now enumerating the world every re-render.

**Comment by Kegsay** [comment]
Right. I have *not yet been able to test this PR* so merge at your own risk. The code LGTM. You based this PR off my `invite-search` PR, which has performance regressions on it. I've backed out those changes such that it no longer blocks this PR being merged to `develop`.

tl;dr - Either you or I need to test this PR for peeking, and if we're happy then we can merge this.

**Comment by Kegsay** [comment]
Re-testing this now that the PR this was based on has landed.

**Comment by Kegsay** [comment]
All seems to work :D LGTM!

**Comment by Kegsay** [comment]
The only feature creepy thing I would request would be a way to be able to peek into a room beyond having to new tab > paste URL. Having a peek/join button on the room directory say would be much joy, so good, wow.

**Comment by ara4n** [comment]
hang on - the intended behaviour is that clicking on a room will always peek by default (if peeking is an option), and you only join if you then hit the join button

**Comment by ara4n** [comment]
also, was scrollback & new-messages working okay on peeked rooms? (i haven't checked if dan's synapse fixes landed)

**Comment by Kegsay** [comment]
> hang on - the intended behaviour is that clicking on a room will always peek by default (if peeking is an option), and you only join if you then hit the join button

This fails on the room directory. This doesn't seem to work when clicking on an invite on the LHS either (my room was configured as `world_readable`, but private).

**Comment by Kegsay** [comment]
> also, was scrollback & new-messages working okay on peeked rooms? (i haven't checked if dan's synapse fixes landed)

Scrollback seemed good, but I am not getting live events now that you mention it.

**Comment by ara4n** [comment]
i'm confused - i can't repro either of this; clicking on a room in the dir correctly peeks at first, and live events work too. going to merge this now, given it's inherited a crapload of CSS refinements from across the board, and we can debug it together tomorrow.

**Comment by Kegsay** [comment]
So I tried this morning `develop` of both react SDK and Vector and:
 - It does peek when you click from the room directory.
 - New messages are received but `/events` is *not woken up* so there is a huge delay before it returns.
 - Invites still do not show any messages (I see the accept/decline prompt, and that's it). Tried with a variety of different configurations of visibility rules / guest access / etc, and nothing helps.
